Add more options for configuring compilers

theotherjimmy opened the pull request:
We added a few more options for configuring compilers in the mbed-os
tools, so I'm adding those options to the readme here. The new options
are:
 - Setting env variables
 - `PATH` detection

Further, I updated the order of the subsections to reflect their
priority. Things earlier in the list configure the compiler in preference
to things later in the list.

Finally, there is one edit for consistency. Previously we had a command
line `blah --global blah blah` and in the text it referred to `-G`.
I changed the command line to be consistent

Reviews/Requests for edits:
 - @AnotherButler 
 - @screamerbg

bridadan commented on README.md:
```diff
@@ -268,16 +268,27 @@ $ mbed remove text-lcd
+There are multiple ways to do this:
+* `mbed_settings.py` file in the root of your program, which is automatically created if it doesn't already exist. 
+* The mbed CLI configuration
+* setting an environment variable
```
Probably make the 's' of "setting" capital here, to make it consistent with the others.

bridadan commented on README.md:
```diff
@@ -268,16 +268,27 @@ $ mbed remove text-lcd
+
+* If you want to use the [ARM Compiler toolchain](https://developer.arm.com/products/software-development-tools/compilers/arm-compiler-5/downloads), set `ARM_PATH` to the *base* directory of your ARM Compiler installation (example: C:\Program Files\ARM\armcc5.06). The recommended version of the ARM Compiler toolchain is 5.06.
+* If you want to use the [GCC ARM Embedded toolchain](https://launchpad.net/gcc-arm-embedded), set `GCC_ARM_PATH` to the *binary* directory of your GCC ARM installation (example: C:\Program Files\GNU Tools ARM Embedded\4.9 2015q2\bin). Use versions 4.9 of GCC ARM Embedded; version 5.0 or any version above might be incompatible with the tools.
+
```
I realize IAR wasn't part of these instructions before, but would you mind adding it here?

AnotherButler commented on README.md:
```diff
@@ -268,16 +268,27 @@ $ mbed remove text-lcd
-* Through the mbed CLI configuration.
-* Via mbed_settings.py file in the root of your program, which is automatically created (if it doesn't already exist). 
+There are multiple ways to do this:
+* `mbed_settings.py` file in the root of your program, which is automatically created if it doesn't already exist. 
```
What automatically creates this? CLI?

AnotherButler commented on README.md:
```diff
@@ -299,14 +310,16 @@ No local configuration is set
+#### Setting environment variable
 
-Edit `mbed_settings.py` to set your toolchain:
+For each of the compilers, there is a corresponding environment variable that is checked for the path to that compiler. The environment variables are as follows:
```
What checks the environment variable for the path to that compiler? The compiler?

AnotherButler commented on README.md:
```diff
@@ -299,14 +310,16 @@ No local configuration is set
 
-Edit `mbed_settings.py` to set your toolchain:
+For each of the compilers, there is a corresponding environment variable that is checked for the path to that compiler. The environment variables are as follows:
+* `MBED_ARM_PATH`: The path to the *base* directory of your ARM Compiler installation. This should be one directory up the directory containing the binaries for `armcc` and friends.
```
Does this mean the directory above or before the directory containing the binaries for `armcc`, or does this mean something else?

theotherjimmy replied:
it's the directory containing the directory containing `armcc`. That's a mouthful. I'm open to suggestions to fix it.

screamerbg commented:
Thanks for aligning functionality and docs @theotherjimmy +1.
See @anotherbutler's comments.

theotherjimmy commented:
@AnotherButler I changed the passive voice you pointed out to active. I also tried to clarify what a base directory is.
@bridadan I fixed the nit. I added IAR.

Could you both review again?

bridadan commented on README.md:
```diff
@@ -268,16 +268,30 @@ $ mbed remove text-lcd
+
+* To use the [ARM Compiler toolchain](https://developer.arm.com/products/software-development-tools/compilers/arm-compiler-5/downloads), set `ARM_PATH` to the *base* directory of your ARM Compiler installation (example: C:\Program Files\ARM\armcc5.06). The recommended version of the ARM Compiler toolchain is 5.06.
+* To use the [GCC ARM Embedded toolchain](https://launchpad.net/gcc-arm-embedded), set `GCC_ARM_PATH` to the *binary* directory of your GCC ARM installation (example: C:\Program Files\GNU Tools ARM Embedded\4.9 2015q2\bin). Use versions 4.9 of GCC ARM Embedded; version 5.0 or any version above might be incompatible with the tools.
+* To use the [IAR EWARM toolhain](https://www.iar.com/iar-embedded-workbench/#!?architecture=ARM), set `IAR_PATH` to the *base* directory of your IAR installation. Use versions 7.80 of IAR EWARM; prior versions might be incompatible with the tools.
```
Last nit! Can you provide an example of the correct directory like above? I always find I forget this myself.

Also: `Use versions 7.80 of IAR EWARM;...`. I think `versions` should be `version`.

AnotherButler commented on README.md:
```diff
@@ -268,16 +268,30 @@ $ mbed remove text-lcd
+* The mbed CLI configuration
+* Setting an environment variable
+* Adding directory of the compiler binary to your PATH
+
```
Please make sure each bullet has a period at the end for consistency.

AnotherButler commented on README.md:
```diff
@@ -268,16 +268,30 @@ $ mbed remove text-lcd
+Edit `mbed_settings.py` to set your toolchain:
+
+* To use the [ARM Compiler toolchain](https://developer.arm.com/products/software-development-tools/compilers/arm-compiler-5/downloads), set `ARM_PATH` to the *base* directory of your ARM Compiler installation (example: C:\Program Files\ARM\armcc5.06). The recommended version of the ARM Compiler toolchain is 5.06.
+* To use the [GCC ARM Embedded toolchain](https://launchpad.net/gcc-arm-embedded), set `GCC_ARM_PATH` to the *binary* directory of your GCC ARM installation (example: C:\Program Files\GNU Tools ARM Embedded\4.9 2015q2\bin). Use versions 4.9 of GCC ARM Embedded; version 5.0 or any version above might be incompatible with the tools.
```
Please change "versions 4.9" to "version 4.9" for agreement.
Please change "or any version above" to "or any more recent version" to fit our style.

AnotherButler commented on README.md:
```diff
@@ -268,16 +268,30 @@ $ mbed remove text-lcd
+* To use the [GCC ARM Embedded toolchain](https://launchpad.net/gcc-arm-embedded), set `GCC_ARM_PATH` to the *binary* directory of your GCC ARM installation (example: C:\Program Files\GNU Tools ARM Embedded\4.9 2015q2\bin). Use versions 4.9 of GCC ARM Embedded; version 5.0 or any version above might be incompatible with the tools.
+* To use the [IAR EWARM toolhain](https://www.iar.com/iar-embedded-workbench/#!?architecture=ARM), set `IAR_PATH` to the *base* directory of your IAR installation. Use versions 7.80 of IAR EWARM; prior versions might be incompatible with the tools.
+
+Since `mbed_settings.py` contains local settings (possibly relevant only to a single OS on a single machine), it should not be checked into version control. 
```
Please change "Since" to "Because" because "since" implies time, and "because" implies causation.
Please change "it should not be checked" to you should not check it" for active voice.

AnotherButler commented on README.md:
```diff
@@ -299,14 +313,16 @@ No local configuration is set
+For each of the compilers, `mbed compile` checks a corresponding environment variable for the compiler's location. The environment variables are as follows:
+* `MBED_ARM_PATH`: The path to the *base* directory of your ARM Compiler installation. This should be the directory containing the directory containing the binaries for `armcc` and friends.
+* `MBED_IAR_PATH`: The path to the *base* directory of your IAR EWARM Compiler installation. This should be one directory containing the directory containing the binaries for `iccarm` and friends.
+* `MBED_GCC_ARM_PATH`: The path to the *binary* directory of your GCC ARM Embedded Compiler installation. This should be the directory containing the binaries for `arm-none-eabi-gcc` and friends.
```
Thanks for clarifying this.

AnotherButler commented on README.md:
```diff
@@ -299,14 +313,16 @@ No local configuration is set
+#### Compiler detection through the `PATH`
 
-As a rule, because `mbed_settings.py` contains local settings (possibly relevant only to a single OS on a single machine), it should not be versioned. 
+If none of the above have been configured, the `mbed compile` command will fall back to checking your `PATH` for an executable that is part of the compiler suite in question. This check is the same as a shell would perform to find the executable on the command line. When it finds the executable it's looking for, it uses the location of that executable as the appropriate path except in the case of ...
```
Please change "none of the above have" to "none of the above has" for agreement.
Please hyphenate "command-line".
Please change "which will not use a path in this case" to "which does not use a path" to eliminate the repetition and for consistent tense.

theotherjimmy commented:
@AnotherButler I re-wrote the patch again to include you second set of suggestions. Could you review again?

I think we should go over this whole file together soon and edit the whole thing.

screamerbg commented:
@AnotherButler Can you please review?

AnotherButler commented:
@screamerbg This looks good to me. theotherjimmy and I put together another PR with some grammar changes for the other sections that this PR did not change, so we've been over the whole document. It looks good.

screamerbg commented:
Thanks @theotherjimmy @AnotherButler
